File: Week_1/P1.java
class Foo {
    int currTurn;

    private void printsWrapper(Runnable printFunction, int turn) throws InterruptedException {
        synchronized(this) {
            while(turn != currTurn) {
                wait();
            }
        }
        printFunction.run();
        synchronized(this) {
            currTurn += 1;
            notifyAll();
        }
    }

    public Foo() {
        this.currTurn = 1;
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printsWrapper(printFirst, 1);
    }

    public void second(Runnable printSecond) throws InterruptedException {
        printsWrapper(printSecond, 2);
    }

    public void third(Runnable printThird) throws InterruptedException {
        printsWrapper(printThird, 3);
    }
}

File: Week_1/P1.cpp
class Foo {

private:
    mutex gMutex;
    condition_variable cv;
    int currTurn;

    void printsWrapper(function<void()> printFunction, int turn) {
        {
            {
                unique_lock<mutex> lk(gMutex);
                cv.wait(lk, [this, turn]{ return (turn == currTurn); });
            }
            printFunction();
            {
                unique_lock<mutex> lk(gMutex);
                currTurn += 1;
            }
            cv.notify_all();
        }
    }

public:
    Foo() {
        this->currTurn = 1;
    }

    void first(function<void()> printFirst) {
        printsWrapper(printFirst, 1);
    }

    void second(function<void()> printSecond) {
        printsWrapper(printSecond, 2);
    }

    void third(function<void()> printThird) {
        printsWrapper(printThird, 3);
    }
};